fix(a11y): improve RWD challege grid accessibility

bbsmooth opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes on my machine.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #45419

<!-- Feel free to add any additional description of changes below this line -->

My second batch of accessibility fixes for this issue. The primary fix is to announce the status of each challenge to screen reader users (they will hear either "Passed" or "Not Passed" after each number). I've also added a jump link so that keyboard users can immediately jump to the next challenge waiting to be completed (potentially saving them a lot of tabbing). I've added the "Challenges" heading above the grid because it seems appropriate here and could help with screen reader navigation. I made a few CSS changes as well, such as using `rem`s to set the width/height on each link so that it will scale better if the user manually increases the font size. Also put a min width/height on them of `24px` in anticipation of WCAG 2.5.8 coming later this year.

I don't consider this 100% complete at the moment as I have some questions that could require changes.

1. The jump link doesn't really look like a link. We could underline it, but I wonder if there is something else we could do to indicate that it's clickable?
2. Do we want to announce Passed/Not Passed, or Completed/Not Completed, or something else to screen reader users?
3. [WCAG 1.4.1](https://www.w3.org/TR/WCAG21/#use-of-color) says that you can't rely on color alone to convey information, so for the challenges that have been completed I increased the font size just slightly (and font weight) in order to have a second visual indicator. I did play around with adding some sort of check mark, but I wasn't happy with it (possibly because I'm not an artist). If anyone has a better idea for identifying the completed challenges I'd love to hear it. Oh, and let me add that there is a very technical loophole that would actually allow us to only use a change in background color, but a lot of accessibility experts consider it illegitimate. Regardless, if we wanted to stick to just a change in background color then it would require increasing the darkness of the blue a little so that it has a 3:1 contrast with the white background.
4. If you don't like the Steps heading above the grid then we could visually hide it.

bbsmooth committed:
fix: improve RWD challege grid accessibility

nhcarrigan commented:
Heya Bruce,

As always, thank you for your continued efforts to improve our accessibility.

> Are we referring to these as "challenges" or "steps" or something else?

Each challenge is referred to as a "step", and the block is referred to as a "practice project".

> The jump link doesn't really look like a link.

We could potentially style this like a button, though I'm not sure how well it would fit in the block card.

> [WCAG 1.4.1](https://www.w3.org/TR/WCAG21/#use-of-color) says that you can't rely on color alone to convey information, so for the challenges that have been completed I increased the font size just slightly (and font weight) in order to have a second visual indicator.

I'm assuming this is for colourblind users? I tried removing the background and the bolded text is pretty hard to notice on its own.

bbsmooth committed:
fix: change wording from challenge to step

bbsmooth committed:
fix: improve keyboard focus indicator on step links

bbsmooth committed:
fix: style jump link as button

bbsmooth committed:
fix: accessible background color for completed steps

bbsmooth commented:
@nhcarrigan Thanks for the feedback/suggestions Naomi. I changed the wording from "Challenges" to "Steps". I styled the jump link as a button but I toned it down just a little (smaller border, less padding). I'm still not sure if this will work? Another thought I had was to perhaps move the link up and to the left of the heading? I'll leave it up to designers around here to make suggestions. I agree with you, the bold on the completed steps is not super obvious, so I've adjusted the background color (for both light/dark mode) so that it meets a 3:1 contrast ratio, which will technically make it accessible. My goal was to make as few changes as possible to get this accessible but perhaps we do need something like a check mark?

bbsmooth committed:
refactor: make steps a nav region

bbsmooth commented:
I've removed the Steps heading and wrapped the steps grid in a `<nav>` with an accessible name for screen reader users ("Steps for [Course Name]"). I think this looks a lot cleaner now.

Still not sure about the text for the jump link. Anybody have a better suggestion?

bbsmooth committed:
refactor: remove duplicate selector

bbsmooth committed:
fix: enhance focus indicator on jump link

bbsmooth committed:
fix: TS type

bbsmooth committed:
refactor: remove unneeded CSS

nhcarrigan commented on client/src/templates/Introduction/components/challenges.tsx:
```diff
@@ -49,40 +52,66 @@ function Challenges({
+            }
+            to={firstIncompleteChallenge.fields.slug}
+          >
+            Jump to next incomplete step
```
A note that we'll need to make this text translatable.

bbsmooth replied:
I like the wording suggestion a lot. I've also added the translation.

bbsmooth replied:
P.S. I also added the translation for the `aria-label` on the `<nav>`. I'm not sure if I did this correctly though? I followed the example in client/src/components/SolutionViewer/ProjectModal.tsx.

naomi-lgbt reviewed (commented):
This is looking great - thanks again for all of your work on this.

Rather than `Jump to next incomplete step`, could we use something less verbose like `Resume project`?

bbsmooth committed:
chore: translate accessible name on nav

bbsmooth committed:
fix: remove h4 from button

emmalearnscode commented on client/src/templates/Introduction/components/challenges.tsx:
```diff
@@ -49,40 +52,71 @@ function Challenges({
+                    +challenge.isCompleted ? 'challenge-completed' : ''
+                  }`}
+                >
+                  <span>{i + 1}</span>
```
This only reads out '1 link' '2 link' ''3 link' and so on which is not clear to a screen reader user that this means lesson/step 1, lesson/step 2, lesson/step 3 etc. I know it's mentioned in the navigation title but that's a high cognitive load to remember by the time you get to link 67.

bbsmooth replied:
@emmalearnscode I've gone back and forth on this. Initially I was going to add a screen reader only "Step" before the number but I was trying to keep the verbosity down and also keep the accessible name staring with the the visible text. To compensate, as you pointed out, I put "Steps" in the accessible name for the nav. But I can definitely see the advantages of adding "Step" to make it clearer. We could go with that and then try to get some feedback from screen reader users.

Concerning the cognitive load issue, if adding "Step" before each number would reduce the cognitive load by making it clear that these are links to steps in the course then could the argument be made that we should also add the course name to these links as well for the same reason?

emmalearnscode replied:
I don't think the course name needs adding to each link. I heard it 3 times already by tabbing first the expandable button, then the resume button and then the navigation and when it's a long title it's pretty verbose. I think the cognitive load for me comes with 'one link' having more than one meaning and I interpret it automatically to mean there is one link but it's not being explained to me, not that the link has the text '1' inside it. I think 'step 1 link' removes the ambiguity without adding too much extra verbosity.

Here's what I hear with NVDA (from the expandable button):
Learn HTML by building a Cat Photo App Not started button collapsed heading level 3

Resume project Learn HTML by building a Cat Photo App link

Steps for Learn HTML by building a Cat Photo App Navigation landmark list with 67 items 1 link

2 link

3 link

bbsmooth replied:
> I interpret it automatically to mean there is one link but it's not being explained to me

Ahh, I see what you are saying. For someone who is completely new to the site I can totally understand that. But I also think most of our users stay for quite a while and would become acquainted with the step navigation list rather quickly and would probably not have this misunderstanding for long. But like I said, I am 50/50 on this, and it is just one extra word, so I'll add it and then if we get any feedback from screen reader users down the road we can address it again.

emmalearnscode commented on client/src/templates/Introduction/components/block.tsx:
```diff
@@ -193,9 +193,9 @@ export class Block extends Component<BlockProps> {
+              <div className='course-title'>
                 {`${isExpanded ? collapseText : expandText}`}
-              </h4>
+              </div>
```
Some extra text for screen readers is needed here. At the moment each one just says 'expand course for 0/39 collapsed' but there's no information about which course it's referring to or what the 0/39 means and there are several of these on the page.

__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix

bbsmooth replied:
I've added the course name to the toggle button and also improved the reading of the course stats for screen readers. I'm still not 100% sure I've done the translations correctly on this part though.

emmalearnscode replied:
It definitely sounds much better now!

__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :ok_hand: This code looks good

EmmaDawsonDev reviewed (commented):
I've made a couple of suggestions based on the PR I made earlier where information could be made clearer for screen reader users.

I like the design of the resume course button and upping the contrast on the completed lessons looks ok but could maybe be iterated on further (I don't think it should block this getting merged for now though). I'd be wary of using a checkmark as it doesn't mean the same thing in all countries (For example in Sweden it means the opposite - that something is not correct).

bbsmooth committed:
fix: add course name to toggle button

bbsmooth committed:
fix: improve accessibility of course stats on toggle button

bbsmooth committed:
fix: prefix step links with sr-only "Step"

bbsmooth committed:
chore: resolve conflict

moT01 reviewed (commented):
This looks real good @bbsmooth 🎉

For reference, here's what it currently looks like: 
![Screen Shot 2022-06-13 at 5 03 28 PM](https://user-images.githubusercontent.com/20648924/173452879-c47a430a-dc5f-4996-a997-d2c4a0031c08.png)

I don't have any issue with the size and weight change on the font for the completed challenges. I played with the inverse, where the uncompleted ones were bigger and bold - it looked about the same. Any thoughts @ahmadabdolsaheb

The only thing I noticed that we probably could change is that the button says and reads 'resume project' if you haven't completed any challenges. I think we should change it to 'Start Project'. Note: the button disappears if you complete the whole project.

I left a comment on one other concern below, I would be fine leaving that if we want.

bbsmooth commented on client/src/templates/Introduction/components/challenges.tsx:
```diff
@@ -49,40 +52,72 @@ function Challenges({
+            }
+            to={firstIncompleteChallenge.fields.slug}
+          >
+            {t('buttons.resume-project')}{' '}
+            {blockTitle && <span className='sr-only'>{blockTitle}</span>}
```
This is purely an accessibility best practice. The visible label on the button ("Resume Project") should be at the beginning of the accessible name used by screen readers. But since there could potentially be several "Resume Project" buttons available in the DOM at once, I've added the project name itself to the end of the accessible name in the `sr-only` text. So we actually don't want this order changed in translations.

moT01 replied:
Interesting. Sounds good to me.

bbsmooth commented on client/src/templates/Introduction/components/block.tsx:
```diff
@@ -193,12 +193,23 @@ export class Block extends Component<BlockProps> {
-                {`${isExpanded ? collapseText : expandText}`}
-              </h4>
+              <div className='course-title'>
+                {`${isExpanded ? collapseText : expandText}`}{' '}
+                <span className='sr-only'>{blockTitle}</span>
```
Same accessibility issues as for the "Resume Project" button.

bbsmooth commented:
> I don't have any issue with the size and weight change on the font for the completed challenges. I played with the inverse, where the uncompleted ones were bigger and bold - it looked about the same.

Since we are using background color that has sufficient color contrast (3:1) to mark the completed steps then we don't technically need to do anything else, such as make the font weight bigger. I just kept it as bold because I thought it would be a little easier to see the numbers over the darker background, although the contrast between the black number and the blue background is high enough that it is accessible regardless.

ahmadabdolsaheb commented:
Thanks for the much needed grid accessibility improvements. Here is the plan for the blues.

Moving forward we will have two CTA's:
1. One to prioritize an action in a group of actions ( Run the tests button )
2. One to prioritize an action in the whole page ( Sign up ). 

For the first CTA, I am going to use a color similar to `#3895ff`.
 
As a result, going bold is a great call here as we could use our current highlight color and still differentiate the completed items in the grid. 

For now let keep the colors as they are, and discuss them more details when we roll out the component library.

ahmadabdolsaheb commented on client/src/templates/Introduction/components/block.tsx:
```diff
@@ -193,12 +193,23 @@ export class Block extends Component<BlockProps> {
+                >{`${completedCount}/${challengesWithCompleted.length}`}</span>
+                <span className='sr-only'>
+                  ,{' '}
+                  {t('learn.challenges-completed', {
+                    completedCount,
+                    totalChallenges: challengesWithCompleted.length
+                  })}
```
This looks great. I was wondering if you have come across a solution for languages with non-english numbers @moT01.

moT01 replied:
I haven't really ever looked into it. A quick search, I found two packages that look okay: https://www.npmjs.com/package/numbo & https://www.npmjs.com/package/n2words. We would probably want to pair up with native speakers to see if converting the numbers in the translations still makes sense. We could just add numbers to the translation object and put it on crowdin - that's probably our best option. I don't think we need to have that many numbers - the longest projects are maybe 150 steps. Then there's also the other miscellaneous numbers across the site, which can already be translated... but they aren't getting translated:
![Screen Shot 2022-06-14 at 7 57 50 AM](https://user-images.githubusercontent.com/20648924/173583272-decd1ab5-49f8-42c9-bd85-1bd3cca04f6f.png)
hmm - don't know. Maybe it's common to use English numbers in Chinese writing or something.

sidemt replied:
To me, a native Japanese speaker and understands a little bit of Chinese, those packages don't look useful here.

At least for Japanese, having a string with interpolation of numbers is the best option. 
We can just use Arabic numbers in Japanese (and it's more natural), but, often the order of the words are different.

Such as the current string: `{{completedCount}} of {{totalChallenges}} challenges completed` is fine.
It can be translated into something like: `{{totalChallenges}}件中{{completedCount}}件完了` and eventually rendered such as `150件中10件完了`

moT01 replied:
Thanks for the input here. I think any changes like this would be outside the scope of the PR. If we want to add this, we should make a new issue. I am going to resolve this - feel free to unresolve if you want to continue the conversation.

ahmadabdolsaheb commented on client/i18n/locales/english/translations.json:
```diff
@@ -72,7 +72,8 @@
     "change-language": "Change Language",
-    "cancel-change": "Cancel Change"
+    "cancel-change": "Cancel Change",
+    "resume-project": "Resume project"
```
There should be a "Start Project" button for projects that have not been started yet.

bbsmooth replied:
Just thinking out loud here. Since we don't display any button when the project is completed, perhaps we shouldn't display any button if the project hasn't been started? A start project button is just taking you to step 1, which is just and extra tab stop to get to step 1 manually.

moT01 replied:
I could go either way on it. I kind of like the "start project" button - it's an obvious thing to click. Otherwise, there's just a list of numbers - pretty safe to assume they will understand that 1 is what to click to start, but it will be even more obvious with the button. I don't mind either way though.

bbsmooth replied:
The people have spoken!

moT01 replied:
This has been added

bbsmooth committed:
fix: add start project button

moT01 commented on client/src/templates/Introduction/components/block.tsx:
```diff
@@ -193,12 +193,23 @@ export class Block extends Component<BlockProps> {
+                  className='map-completed-count'
+                >{`${completedCount}/${challengesWithCompleted.length}`}</span>
+                <span className='sr-only'>
+                  ,{' '}
```
This one does add the comma and space to voice over, so I see the need for this one.

bbsmooth replied:
Ironically, this is the only space that the formatter didn't put back in because it's not at the end of the line.

bbsmooth replied:
Space added back by formatter during commit.

moT01 commented on client/src/templates/Introduction/components/challenges.tsx:
```diff
@@ -49,40 +52,78 @@ function Challenges({
+            to={firstIncompleteChallenge.fields.slug}
+          >
+            {!isChallengeStarted
+              ? t('buttons.start-project')
+              : t('buttons.resume-project')}{' '}
```
```suggestion
              ? t('buttons.start-project')
              : t('buttons.resume-project')}{' '}
```

This one didn't seem to make a difference in voice over either. The space is there either way. Maybe other screen readers don't add the space - not sure. If we're keeping these, wouldn't we want to add the space in the start project button too? `? t('buttons.start-project'){' '}`

bbsmooth replied:
Again, space added by formatter during commit.

moT01 reviewed (dismissed):
LGTM @bbsmooth 🎉 

I left a few comments on one curiosity I had. I would be fine merging it as-is.

bbsmooth committed:
refactor: remove unnecessary spaces

bbsmooth commented on client/src/templates/Introduction/components/block.tsx:
```diff
@@ -193,12 +193,23 @@ export class Block extends Component<BlockProps> {
-                {`${isExpanded ? collapseText : expandText}`}
-              </h4>
+              <div className='course-title'>
+                {`${isExpanded ? collapseText : expandText}`}{' '}
```
@moT01 Mystery solved. I actually didn't add these spaces manually. The formatter is adding them when you do a commit. I just proved it. I removed the particular space and when I did the commit the formatter added it back in there. Must have something to do with the necessary space being at the end of the line.

bbsmooth commented:
I was able to remove one unnecessary space because it wasn't at the end of the line so the formatter didn't add it back in.

moT01 reviewed (dismissed):
LGTM 🎉

ahmaxed committed:
fix: clean up styles

ahmadabdolsaheb commented on client/src/templates/Introduction/intro.css:
```diff
@@ -44,11 +44,23 @@ a.cert-tag:active {
 }
 
+.block-grid .challenge-jump-link {
+  margin: 0 25px 25px 25px;
```
```suggestion
  margin: 0 25px 25px;
```

ahmaxed committed:
Update client/src/templates/Introduction/intro.css

ahmadabdolsaheb commented:
Yes, I would need a little time to work on our color system before we introduce new shades of blue. Thank you for your patience.
